Add design proposal for supporting dependencies between build artifacts

gsquared94 opened the pull request:
Added a design proposal for supporting build artifact dependencies.

tejal29 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,223 @@
+## Build logs reporting
+
+There are two quirks in the way the current `InParallel` build controller reports the build logs.
+
```
```suggestion
Current Implementation
```

Its been long since i touched this code and first thought the below snippet it proposed code change

gsquared94 replied:
added

tejal29 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,223 @@
+
+This will prove misleading after the introduction of artifact dependencies since we can have out of order artifact definitions in the skaffold config which with the current reporting strategy would appear to be building in the wrong order, and also build failures due to failed required artifact builds won't be immediately apparant.
+
+So we introduce a new `BuildLogger` that is both order aware while at the same time does not interleave log messages due to some builds running concurrently. Additionally it concats all build failure messeges and reports all of them.
```
Do you have the details of this scoped somewhere?
I would suggest, we first implement
1. Artifact dependency in sequential order. Meaning a single artifact will be built at a time. Keep in mind, we design for both running in sequence an in parallel. However, in-sequence builds will be faster to land.
2. Design UX for running builds in parallel
3. Switch to running builds concurrently.

tejal29 replied:
It makes sense to design the UI and then add functionality execute builds concurrently to make sure we add features with good UX.

WDYT?

tejal29 replied:
Add a section of "Future work" and add "Re-desing UX for concurrent builds"

gsquared94 replied:
Added your suggestions

tejal29 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,223 @@
+- Cyclic references among artifacts.
+  - We cannot have image `A` depend on image `B` depend on image `C` depend on image `A`.
+  - We run a simple depth first search cycle detection algorithm treating our `Artifact` slice like a directed graph- image `A` depending on image `B` implies a directed edge from `A` to `B`.
+- Unique artifact aliases.
```
Few things to consider, 
1. Can a base artifact be deployed?
- Is yes, then a base artifact can have sync rules and the would directly sync in the deployed container.
- if not, 
   - What happens to the sync rules defined for the artifact?
       a. Should they be combined with the sync rules defined in the dependency artifact?
            - If yes, what is the conflict resolution if they overlap.

In my opinion, a base artifact most of the times is a shared library code and not a binary. 
We can assume #1 and disallow any sync rules defined for a base artifact.
Add a section of assumptions and list this into assumptions and see what users have use cases where the deploy a base image as a container or they want to inherit sync rules from base image.
WDYT?

2. Since, this "alias" syntax is only applicable for docker builder, `requires` should be allowed only on `DockerArtifact`.

gsquared94 replied:
> 1.
Yes base artifact can be deployed and it can define its own sync rules. When considering it as a dependency however we are just concerned with the image that it builds and ignore any sync rules. I don't think that there's a reliable way to merge base sync rules since we don't really know how the base image is gonna be used in the dependent artifact's build. WDYT?
cc. @briandealwis 
> 2.
I added that to the document scope. We can add a validation to check this and remove it later as we make progress with other builders.

gsquared94 replied:
https://github.com/GoogleContainerTools/skaffold/issues/3247 and https://github.com/GoogleContainerTools/skaffold/issues/889 are two reported instances. I don't think there are any reported cases involving BASE artifacts that have sync definitions. 

>Consider a case where shared are assets are defined in a base image. Ideally, sync rules for this asset would be part of the artifact. If they are ignored, they would end up duplicating in all images where they are used.

This example definitely helps me understand the confusion that you're talking about. Maybe I'm wrong about this: "don't think that there's a reliable way to merge base sync rules"

We can probably, with some effort parsing the Dockerfile, figure out if the derived artifact is copying selective files like in a multistage build or just using it as a base image and try to create matching sync rules in the derived artifact? Anyway seems like a good followup candidate; we can start off by forcing a rebuild and ignore being able to just sync the files. If there are enough objections to that approach we can try to optimize later? WDYT?

briandealwis replied:
A very thought-provoking question, @tejal29.  Syncing and base images has deep implications, especially since we now provide a SKAFFOLD_RUN_MODE that a Dockerfile could use to alter its behaviour, such as to run a file-watcher to process sync'd files.  Those file-watchers may not be invoked in child images.

For example, consider a base image that sets up a watcher to compiles a set of .java files to .jar files (e.g., `gradle --continuous`).  This image is used by a set of child images that run different jars with different entrypoints.  If we inherit sync rules from the base, then there's a strong possibility that the sync'd files are ignored.

It seems to me that images used as dependencies may need to be built with SKAFFOLD_RUN_MODE=build, and any attending changes to build-flags.

I'll note that syncing is an optimization — we should always be able to fall-back to rebuilding images.

nkubala replied:
trying to merge sync rules from a base image seems like it will be difficult to get right in practice - there are lots of scenarios I can imagine where sync rules from the base overwrite the pre-existing sync rules from a dependent image in a way that's unexpected for the user.

for the first implementation I think we should just ignore sync rules from a base image when it isn't only being deployed on its own. we could consider allowing annotations on the base image to tell skaffold to honor its sync rules, but this should be a workaround and not the default.

gsquared94 commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,223 @@
+}
+```
+
+Each artifact has a channel that it closes once it completes building (either success or failure) by calling `markComplete`. This notifies *all* listeners waiting for this artifact.
````
I was initially thinking about handling the failure separately. But thinking about your comment made me realize that I can handle it here it self. Modified the suggested code. PTAL.

gsquared94 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,223 @@
+   - image: leeroy-app
+     requires:
+       - image: simple-go-app
+         alias: BASE
```
done

gsquared94 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,223 @@
+
+[InSequence](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/sequence.go) and [InParallel](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go) are two build controllers for deciding how to schedule the run of multiple builds together. `InSequence` runs all b ...
+
+After introducing inter-artifact dependencies we'll need to run the builds in a topologically sorted order. One way to do this is to run any popular topologically sorting algorithm on the `Artifact` slice, treating it as a directed graph. However we can get a simpler implementation at the expense of a few additional `goroutines`.
```
done

tejal29 commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,246 @@
+We scope our initial implementation to support artifact dependencies only for docker artifacts. This means that docker artifacts can define dependencies on any other artifact types. It doesn't limit another artifact type from being used as a dependency.
+## Config schema
+
+We introduce an `ArtifactDependency` slice within `Artifact` in `config.go`
```
With the new scope, can this part of just the `DockerArtifact` struct? If we don't plan to support modules for other builders in near future, it will make sense to just add it to `DockerArtifact`

gsquared94 replied:
I want to leave this inside `Artifact` since this will allow use to have all the code plumbing done for all the builders for when we do implement it for them. I think `Jib` and `Kaniko` also support base images (maybe multistage also? don't know that) and it should be possible to implement this feature for them too.

tejal29 replied:
> Artifact dependencies apply to other builder types, and even between builder types too.

Agreed. The current design only supports implementing it for dockerfile builder by relying on the `ARG` command in the Dockerfile and passing currently built image to override the arg.
For other builders, jib, buildpacks, etc, design & implementation plan is needed.

The only reason, i prefer moving this to `DockerArtifact` is having at artifact level will allow users (especially ide users) to configure "artifact dependency tag" on *any* artifact type. 
Letting users configure this for any artifact type but spitting out an error later isn't very user friendly *especially* if there is a way to avoid it.


>I want to leave this inside Artifact since this will allow use to have all the code plumbing done for all the builders for when we do implement it for them. I think Jib and Kaniko also support base images (maybe multistage also? don't know that) and it should be possible to implement this feature for them too.

Agreed. We can add the field to individual types as we implement them and move it to global once all artifacts have the support.

tejal29 replied:
re
> I disagree with scoping this to docker-artifacts. Artifact dependencies apply to other builder types, and even between builder types too.

Do you mean we should address other builds in this DD as well ?

nkubala replied:
as discussed, let's broaden the scope of this design to all builders. the implementation can start with docker artifacts since it's probably the most complicated.

briandealwis replied:
> @tejal29: We can add the field to individual types as we implement them and move it to global once all artifacts have the support.

💡!  Now I understand what you mean.  +1 to putting this on DockerArtifact.

tejal29 reviewed (commented):
Thanks for the explanation. Few other follows up can be discussed in a team meeting setting. 

just 1 qq.  (pardon me if this was clarified in the design document linked)
```
build:
  tagPolicy:
    gitCommit: {}
  - image:simple-go-app
    ..
   - image: leeroy-app
     requires:
       - image: simple-go-app
         alias: BASE
```
In this case with the tagging strategy, for the dependent docker build arg BASE will replaced with fully qualified image name right.
e.g. if default-repo is set to "gcr.io/X`
```
BASE = "gcr.io/X/simple-go-app@<TAG>
```
Maybe you could add this in section called "Docker Builder Changes" if not in the parent design document.

gsquared94 commented:
> Thanks for the explanation. Few other follows up can be discussed in a team meeting setting.
> 
> just 1 qq. (pardon me if this was clarified in the design document linked)
> 
> ```
> build:
>   tagPolicy:
>     gitCommit: {}
>   - image:simple-go-app
>     ..
>    - image: leeroy-app
>      requires:
>        - image: simple-go-app
>          alias: BASE
> ```
> 
> In this case with the tagging strategy, for the dependent docker build arg BASE will replaced with fully qualified image name right.
> e.g. if default-repo is set to "gcr.io/X`
> 
> ```
> BASE = "gcr.io/X/simple-go-app@<TAG>
> ```
> 
> Maybe you could add this in section called "Docker Builder Changes" if not in the parent design document.

Added section `Referencing dependencies inside docker builder` and also added `future work` with initial idea about how the other builder could support it.

briandealwis reviewed (commented):
Should have submitted this as individual comments rather than as a review

nkubala commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,253 @@
+
+There are two quirks in this: 
+- It reports in the order of artifacts in the `Artifact` slice instead of the actual order in which they get built. 
+- It only reports a single artifact build failure even though there could have been multiple failures.
```
https://github.com/GoogleContainerTools/skaffold/pull/2077 and https://github.com/GoogleContainerTools/skaffold/pull/2000 :)

nkubala commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,253 @@
+- It reports in the order of artifacts in the `Artifact` slice instead of the actual order in which they get built. 
+- It only reports a single artifact build failure even though there could have been multiple failures.
+
+This will prove misleading after the introduction of artifact dependencies since we can have out of order artifact definitions in the skaffold config which with the current reporting strategy would appear to be building in the wrong order, and also build failures due to failed required artifact builds won't be immediately apparant.
```
```suggestion
This will prove misleading after the introduction of artifact dependencies since we can have out of order artifact definitions in the skaffold config which with the current reporting strategy would appear to be building in the wrong order, and also build failures due to failed required artifact builds won't be immediately apparent.
```

nkubala reviewed (commented):
this looks really good, I think it's close to done. @briandealwis and @tejal29 any final feedback?

tejal29 reviewed (approved):
LGMT to me except for Buildpacks. I will defer to Brain.

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,356 @@
+       - image: simple-go-app
+         alias: BASE
+```
+Alias is a token that will be replaced with the image reference in the builder definition files. If no value is provided for `alias` then it defaults to the value of `image`.
````
Some builders may not support this as they may have restrictions on alias names.  For example, the docker and custom builders require aliases to be valid environment variable names and so do not allow `/` or `.`.

gsquared94 replied:
if `alias` is missing and `image` name is not in the correct format, then it'll fail the validation immediately and the user can add an explicit alias then.

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+
+### Docker builder
+
+The `docker` builder will use the `alias` of an `ArtifactDependency` as a build argument key.
```
Define what is a valid alias (e.g., `[A-Za-z_][A-Za-z0-9_]*`).

gsquared94 replied:
added to validation section

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+
+### Custom builder
+
+The `custom` builder will be supplied each `ArtifactDependency`'s `alias` and image reference as environment variables keyed on `alias`. So they can be easily referenced in user-defined build definitions.
```
Define what is a valid alias as per the Docker builder

gsquared94 replied:
all aliases have the same restrictions since they are either used as environment variable keys or build arg keys or template variables. So I've added it in the validation section above.

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,356 @@
+    buildpacks:
+      builder: "{{ .BUILDER_IMAGE }}"
+      runImage: "{{ .RUN_IMAGE }}"
+    requires:
+      - image: builder-image
+        alias: BUILDER_IMAGE
+      - image: run-image
+        alias: RUN_IMAGE
+```
+
+If there are any additional images in the `required` section it only enforces that they get built prior to the current image. However, the buildpacks builder cannot really reference them in any other way.
````
I've been thinking of our chat yesterday and what [@tejal29 suggested above](https://github.com/GoogleContainerTools/skaffold/pull/4794/files#r491234437), and I now think that Buildpacks shouldn't have these additional `requires` block, since the `builder` and `runImage` are sufficient and encode implicit dependencies.  We're repeating ourselves in the example above, and it doesn't seem to buy us anything.

Let's see if we need build ordering (implicit or explicit) and encode it then.

gsquared94 replied:
It'd be useful to think about this design document with the context that my primary objective is to implement `modules` piece by piece. 

If we are saying that only the docker or custom builder can defined required modules (which could be bundles of multiple builds) then it makes sense to only implement `required` sections for those builders only. Otherwise it'll just complicate the implementation down the line; this feature by itself might only live for a couple of weeks. 

The `required` clause will migrate to the `module` struct soon enough which will be one level above `build`. As such it doesn't seem worthwhile to push it even lower into individual builder structs only to have to refactor again.

With `modules` and `hooks` skaffold is slowly transforming into a general orchestration framework. I think it's a worthwhile tradeoff to have some redundancies in schema to get conformity across whatever builders and deployers and hooks any random module might define.

gsquared94 replied:
TLDR; for this buildpacks example what would the schema look like if `builder-image` was in a skaffold module in another repo?

briandealwis replied:
Good point on the modules.  I'm hoping we don't use Go templates for this; I'd rewrite your example as:
```
build:
  artifacts:
  - image: builder-image
  - image: skaffold-buildpacks
    buildpacks:
      builder: builder-image
    requires:
      - image: builder-image
        module: 
           path: ../builder/ 
```

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+    jib:
+      type: maven
+      args: 
+      - -DbaseImage=docker://{{ .BASE_IMAGE }}
```
This can be [`jib.from.image`](https://github.com/GoogleContainerTools/jib/blob/2085d2192adf2f86fcbfcb1e62f0ed113349426f/jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/PropertyNames.java#L22) too.

You're not showing the definition of BASE_IMAGE?

gsquared94 replied:
fixed

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,356 @@
+where `baseImage` is added as a property to the `pom.xml`
+
+```xml
+<properties>
+  <baseImage>image</baseImage>
+</properties>
+...
````
This bit is unnecessary.

gsquared94 replied:
fixed

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+
+### Bazel builder
+
+The bazel builder doesn't support referencing images directly. Also, it natively supports setting up nested builds. We allow defining `required` artifacts even though they can't be referenced by the builder. We do this to give the user a way of ordering these builds; with the future work around [Skaffold Hooks](https://github.com/GoogleContainerTools/skaffold/issues/1441) there might be some usec ...
```
I'm on the fence about allowing `required` to provide implicit ordering.

gsquared94 replied:
PTAL at comment [above](https://github.com/GoogleContainerTools/skaffold/pull/4794/files#r494065590)

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+
+Skaffold currently allows specifying the `concurrency` property in `build` which affects how many builds can be running at the same time. However it doesn't address the issue of certain builders (`jib` and `bazel`) not being safe for multiple concurrent runs against the same workspace or context. We can fix this also since we are reworking the build controller anyways. 
+
+We define a concept of lease on workspaces by preprocessing the list of artifacts. Each builder tries to acquire a lease on the context/workspace prior to starting the build. Only workspaces associated with concurrency-safe builders allot multiple leases, otherwise it assigns one lease at a time.
```
What happens if we have several artifacts of different builder-type but same context?

gsquared94 replied:
> `Only workspaces associated with concurrency-safe builders allot multiple leases`

So even if one builder is concurrency unsafe, it'll be single leases for that context.

briandealwis replied:
Sounds like we need a builder interface (rather than just the `ArtifactBuilder` function signature).  Can you sketch out what this interface would look like?

gsquared94 replied:
Added

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+		case <-ctx.Done():
+		case <-depStatus.failure:
+			return fmt.Errorf("failed to build required artifact: %q", depStatus.imageName)
+		case <-dep:
```
What's `dep` here?

gsquared94 replied:
`dep` here is a typo! 🤦🏽‍♂️

gsquared94 replied:
fixed

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,356 @@
+
+[hash.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/cache/hash.go) provides the `getHashForArtifact` function that needs to recursively be called for each of its dependencies and all those values aggregated together would be the hashcode for the artifact. This would ensure that for a cache hit all the cascading dependencies a ...
+
+> Note: Builders like docker or buildpacks will have their dependencies provided as build args and environment variables but these specific values are omitted from hash calculation here as they are already accounted for above.
```
```suggestion
> Note: Builders like docker or custom will have their dependencies provided as build args and environment variables but these specific values are omitted from hash calculation here as they are already accounted for above.
```

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,361 @@
+- Unique artifact aliases.
+  - We ensure that within *each* artifact dependency slice the aliases are unique. 
+- Valid aliases
+  - Since aliases are used as environment variable keys or template keys we validate that they match the regex `[a-zA-Z_][a-zA-Z0-9_]*`
```
I think we should move this requirement into the applicable builders.  Just in case some builders may find it useful to be able to alias image-refs directly, like a rewrite rule.

gsquared94 replied:
I think adding it as a config validation helps to show the error early and immediately. I've changed it to validate only for docker and custom builders where we are using the alias with naming restrictions.

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,361 @@
+          alias: BASE
+ ```
+
+Here `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` is passed as a `--build-arg` (or a buildKit parameter) when building `leeroy-app`.
````
```suggestion
Here the alias is used to populate a build arg `BASE=gcr.io/X/simple-go-app:<tag>@sha:<sha>` is passed as a `--build-arg` (or a buildKit parameter) when building `leeroy-app`.
```

gsquared94 replied:
done

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,361 @@
+}
+```
+
+This necessitates all multi-artifact builder implementations to also accept a slice of already built artifacts' information. All single artifact builders require an `ArtifactResolver` that can provide the required artifacts.
````
It would be helpful to explain why multi-artifact builders need a list of already-built artifacts (vs an ArtifactResolver)

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,361 @@
+[InSequence](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/sequence.go) and [InParallel](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go) are two build controllers for deciding how to schedule the run of multiple builds together. `InSequence` runs all b ...
+
+After introducing inter-artifact dependencies we'll need to run the builds in a topologically sorted order.
+We introduce a new controller `scheduler.go` and deprecate `sequence.go` and `parallel.go`. Here we model the `Artifact` slice graph using a set of `go channels` to achieve the topologically sorted build order.
```
Are we deprecating them?  You're planning to remove them, right?

gsquared94 replied:
yes, deleting them

gsquared94 replied:
fixed

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,361 @@
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
+```
+The first describes a builder for a list of artifacts. There are a few implementations like `cache`, `local` and `cluster` builders.
````
Maybe we should rename this from `Builder` to `Platform` or something.

gsquared94 replied:
discussed offline. Resolving.

loosebazooka commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,366 @@
+  - image: test-jib-maven
+    jib:
+      type: maven
+      baseImage: base-image
```
If the base image is always published first, I don't think this is a problem. However if the base image is pushed to a docker daemon and not a registry, you must tell jib where to look. (https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#setting-the-base-image)

gsquared94 replied:
yes, we account for that. Since Skaffold knows about the dependent image being available in a local docker daemon or a registry it'll put the appropriate prefix.

loosebazooka replied:
Sounds good, LGTM on the jib section.

loosebazooka commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,366 @@
+  - image: test-jib-maven
+    jib:
+      type: maven
+      baseImage: base-image
```
We have some new code that allows the user to specify the target platform that might conflict with some of the ideas here (for instance if the original image is a manifest list and the user is selecting a specific platform). However, I don't think that's necessarily within scope of this doc and I'm not even sure it will come up for a 99% of users. Just wanted to make a record of it in case we need to handle it later.

gsquared94 replied:
thanks for that info. We can take that up as a follow up work if it ever comes to it (since 99% of users should be unaffected)

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,366 @@
+
+```go
+type Lease interface {
+  Acquire() error
+  Release() error
````
Do we need to expose the lease itself?  Maybe the LeaseProvider could provide a release function
```
release, err := provider.Acquire(artifact)
defer release()
```
?

gsquared94 replied:
👍

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
```diff
@@ -0,0 +1,366 @@
+	return dag
+}
+
+func (dag *artifactDAG) allDependents(artifact *latest.Artifact) []*latest.Artifact {
```
Maybe rename to `dependents` since it's not the transitive closure.

gsquared94 replied:
done

briandealwis commented on docs/design_proposals/artifact-dependencies.md:
````diff
@@ -0,0 +1,366 @@
+}
+```
+
+This lets us query for _all artifacts that require a given artifact_.
````
Maybe mention the transitive closure.

gsquared94 replied:
done

briandealwis reviewed (approved):
LGTM.
